scripts/nb-tester/qiskit_docs_notebook_tester/config.py

```diff
@@ -0,0 +1,448 @@
+# This code is a Qiskit project.
+#
+# (C) Copyright IBM 2023.
+#
+# This code is licensed under the Apache License, Version 2.0. You may obtain a
+# copy of this license in the LICENSE file in the root directory of this source
+# tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this copyright
+# notice, and modified files need to carry a notice indicating that they have
+# been altered from the originals.
+
+from __future__ import annotations
+
+
+import argparse
+from textwrap import dedent
+from dataclasses import dataclass
+from pathlib import Path
+import tomli
+from typing import Iterator, Literal, Callable
+
+import nbformat
+from squeaky import clean_notebook
+
+# We always run the following code in the kernel before running the notebook
+PRE_EXECUTE_CODE = """\
+# Import with underscores to avoid interfering with user-facing code.
+from matplotlib import set_loglevel as _set_mpl_loglevel
+
+# See https://github.com/matplotlib/matplotlib/issues/23326#issuecomment-1164772708
+_set_mpl_loglevel("critical")
+"""
+
+
+def get_package_versions():
+    requirements_file = Path("scripts/nb-tester/requirements.txt").read_text()
+    package_versions = "\n".join(
+        line for line in requirements_file.split("\n") if not line.startswith("#")
+    )
+    return package_versions.strip()
+
+
+# This markdown replaces cells with tag 'version-info'
+VERSION_INFO = f"""\
+<details>
+<summary><b>Package versions</b></summary>
+
+The code on this page was developed using the following requirements.
+We recommend using these versions or newer.
+
+```
+{get_package_versions()}
+```
+</details>
+"""
+
+
+def post_process_notebook(nb: nbformat.NotebookNode) -> nbformat.NotebookNode:
+    for cell in nb.cells:
+        # Remove execution metadata to avoid noisy diffs.
+        cell.metadata.pop("execution", None)
+        if "version-info" in cell.metadata.get("tags", []):
+            cell.source = VERSION_INFO
+
+    nb, _ = clean_notebook(nb)
+    return nb
+
+
+@dataclass
+class Result:
+    ok: bool
+    reason: str | None = None
+
+    def __bool__(self) -> bool:
+        return self.ok
```

Comment:
Aside from separating the code and introducing the `NotebookJob` interface, this is the only other change. It's useful to know "why" a bool is set so we can give feedback to users. This was introduced for the `write` option but I've switched most functions that return a `bool` to return a `Result` object.